src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,370 @@
+    if (max_data_size < sizeof(void *)) {
+        SOL_WRN("IPM max data size < sizeof(void *)");
+        return -EINVAL;
+    }
+
+    ipm_max_id = ipm_max_id_val_get(ipm_sender);
+
+    ipm_register_callback(ipm_receiver, ipm_receiver_isr_cb, NULL);
+    r = ipm_set_enabled(ipm_receiver, 1);
+    if (r != 0) {
+        SOL_WRN("Could not enable IPM receiver: %s", sol_util_strerrora(r));
+        return -r;
+    }
+
+    initialised = true;
+
+    return 0;
+}
+
+SOL_API int
+sol_ipm_set_receiver(uint32_t id,
+    void (*receive_cb)(void *cb_data, uint32_t id, struct sol_blob *message),
+    const void *data)
+{
+    int idx;
+    struct sol_ipm_receiver_entry *receiver;
+
+    if (!initialised) {
+        int r = init();
+        SOL_INT_CHECK(r, != 0, r);
+    }
+
+    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
+    SOL_INT_CHECK(id, == 0, -EINVAL);
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
+        if (receiver->id == id) {
+            if (receive_cb) {
+                receiver->receive_cb = receive_cb;
+                return 0;
+            } else {
+                return sol_vector_del(&ipm_receivers, idx);
+            }
+        }
+    }
+
+    /* NULL receive_cb means it's trying to remove a receiver, but none was found */
+    SOL_NULL_CHECK(receive_cb, -EEXIST);
+
+    receiver = sol_vector_append(&ipm_receivers);
+    SOL_NULL_CHECK_MSG(receiver, -ENOMEM, "Could not add receiver to id %" PRIu32,
+        id);
+
+    receiver->id = id;
+    receiver->receive_cb = receive_cb;
+    receiver->data = data;
+
+    return 0;
+}
+
+SOL_API int
+sol_ipm_send(uint32_t id, struct sol_blob *message)
+{
+    struct sol_blob_ipm_sent *blob;
+    int r;
+
+    if (!initialised) {
+        r = init();
+        SOL_INT_CHECK(r, != 0, r);
+    }
+
+    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
+    SOL_INT_CHECK(id, == 0, -EINVAL);
+    SOL_NULL_CHECK(message, -EINVAL);
+
+    blob = calloc(1, sizeof(struct sol_blob_ipm_sent));
+    SOL_NULL_CHECK(blob, -ENOMEM);
+
+    sol_blob_setup(&blob->base, SOL_BLOB_TYPE_NO_FREE, message->mem, message->size);
+    sol_blob_set_parent(&blob->base, message);
+    blob->id = id;
+
+    r = ipm_send(ipm_sender, 1, id, &blob, sizeof(uintptr_t));
+    SOL_INT_CHECK_GOTO(r, > 0, fail);
+
+    return 0;
+
+fail:
+    sol_blob_unref(&blob->base);
+    return -r;
+}
+
+SOL_API int
+sol_ipm_set_consumed_callback(uint32_t id,
+    void (*message_consumed_cb)(void *data, uint32_t id, struct sol_blob *message),
+    const void *data)
+{
+    int idx;
+    struct sol_ipm_consumed_handler *handler;
+
+    if (!initialised) {
+        int r = init();
+        SOL_INT_CHECK(r, != 0, r);
+    }
+
+    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
+    SOL_INT_CHECK(id, == 0, -EINVAL);
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_consumed_handlers, handler, idx) {
+        if (handler->id == id && message_consumed_cb)
+            return -EEXIST;
+
+        return sol_vector_del(&ipm_consumed_handlers, idx);
```

Comment:
On 09-05-2016 16:59, Gustavo Sverzut Barbieri wrote:
> In src/lib/io/sol-ipm.c
> <https://github.com/solettaproject/soletta/pull/1996#discussion_r62562414>:
>
>> +    int idx;
>> +    struct sol_ipm_consumed_handler *handler;
>> +
>> +    if (!initialised) {
>> +        int r = init();
>> +        SOL_INT_CHECK(r, != 0, r);
>> +    }
>> +
>> +    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
>> +    SOL_INT_CHECK(id, == 0, -EINVAL);
>> +
>> +    SOL_VECTOR_FOREACH_IDX (&ipm_consumed_handlers, handler, idx) {
>> +        if (handler->id == id && message_consumed_cb)
>> +            return -EEXIST;
>> +
>> +        return sol_vector_del(&ipm_consumed_handlers, idx);
>
> what? if you set a callback == NULL it will delete all other ids :-)
> |handler->id == id| false, then |return sol_vector_del(, idx)|

It doesn't delete all,  only first, as it's a return =P
-ETOOMUCHDORGAS

>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/solettaproject/soletta/pull/1996/files/64ac7ea3f07d3a26a3d264869f28dc376ba64cc9#r62562414>
>